mutiny-core/src/lib.rs

```diff
@@ -1369,39 +1429,94 @@ impl<S: MutinyStorage> MutinyWallet<S> {
         Ok(MutinyBalance::new(ln_balance, federation_balance))
     }
 
+    fn get_invoice_internal(
+        &self,
+        key: &str,
+        inbound: bool,
+        labels_map: &HashMap<Bolt11Invoice, Vec<String>>,
+    ) -> Result<Option<MutinyInvoice>, MutinyError> {
+        if let Some(info) = self.storage.get_data::<PaymentInfo>(key)? {
+            let labels = match info.bolt11.clone() {
+                None => vec![],
+                Some(i) => labels_map.get(&i).cloned().unwrap_or_default(),
+            };
+            let prefix = match inbound {
+                true => PAYMENT_INBOUND_PREFIX_KEY,
+                false => PAYMENT_OUTBOUND_PREFIX_KEY,
+            };
+            let payment_hash_str = key
+                .trim_start_matches(prefix)
+                .splitn(2, '_') // To support the old format that had `_{node_id}` at the end
+                .collect::<Vec<&str>>()[0];
+            let hash: [u8; 32] = FromHex::from_hex(payment_hash_str)?;
+
+            return MutinyInvoice::from(info, PaymentHash(hash), inbound, labels).map(Some);
+        };
+
+        Ok(None)
+    }
+
     /// Get the sorted activity list for lightning payments, channels, and txs.
-    pub async fn get_activity(&self) -> Result<Vec<ActivityItem>, MutinyError> {
-        // Get activity for lightning invoices
-        let lightning = self
-            .list_invoices()
-            .map_err(|e| {
-                log_warn!(self.logger, "Failed to get lightning activity: {e}");
-                e
-            })
-            .unwrap_or_default();
+    pub fn get_activity(
+        &self,
+        limit: Option<usize>,
+        offset: Option<usize>,
+    ) -> Result<Vec<ActivityItem>, MutinyError> {
+        let index = {
+            let index = self.storage.activity_index();
+            let vec = index.try_read()?.clone().into_iter().collect_vec();
+
+            match (limit, offset) {
+                (None, None) => vec,
+                (Some(limit), Some(offset)) => {
+                    let start = offset;
+                    let end = start.saturating_add(limit).min(vec.len());
+                    vec[start..end].to_vec()
+                }
+                (None, Some(offset)) => vec[offset..].to_vec(),
+                (Some(limit), None) => vec[..limit].to_vec(),
+            }
```

Comment:
```suggestion
            let (start, end) = match (limit, offset) {
                (None, None) => (0, vec.len()),
                (Some(limit), Some(offset)) => {
                    let start = offset.min(vec.len());
                    let end = start.saturating_add(limit).min(vec.len());
                    (start, end)
                }
                (None, Some(offset)) => (offset.min(vec.len()), vec.len()),
                (Some(limit), None) => (0, limit.min(vec.len())),
            };

            vec[start..end].to_vec()
```

I think you can make this easier to read and safer by limiting the scope of what happens in the match. Just make it return the start and end and only need to call `to_vec()` in one place.